### PR TITLE
Handle server requests

### PR DESCRIPTION
**Pull request by heitor-lassarote**
This PR adds support for handling server requests in linol.

This PR is based on the `wip-lsp-1.14` branch, and as such, that branch should be merged first.

### CODE REVIEW

**Comment by heitor-lassarote** [comment]
@c-cube, @Sorokin-Anton, requesting your reviews on this. :)

**Comment by c-cube** [comment]
This looks great, at first glance! There's even handling of batch requests! :grin:

**Comment by heitor-lassarote** [comment]
Ohhh, hold on! Actually, it seems batch request handling is not entirely correct, as the server should _return an array_ of responses rather than individual responses. I will fix that and send a new commit with the patch soon! Sorry about this.

As for the remainder: I've tested this on the LIGO language server for the purpose of handling dynamic configuration changes, and it seems to be working well on this matter.

**Review by c-cube — APPROVED** [review]
Looks quite nice, thank you for the refactoring too (and the descriptive commit msg)

**Comment by c-cube** [comment]
going to give your colleague a bit of time to comment before I merge :)

File: src/server.ml [hunk 1]
```diff
@@ -109,7 +124,15 @@ module Make (IO : IO) = struct
...
       method send_notification (n : Lsp.Server_notification.t) = notify_back n
-      (** Send a notification (general purpose method) *)
+      (** Send a notification from the server to the client (general purpose method) *)
+
+      method send_request
+          : 'from_server.
+            'from_server Lsp.Server_request.t ->
+            (('from_server, Jsonrpc.Response.Error.t) result -> unit IO.t) ->
+            Req_id.t IO.t =
+        fun r h -> server_request @@ Request_and_handler (r, h)
```
Comment (Sorokin-Anton) [inline src/server.ml:134]: Maybe call field `send_server_request`?

**Review by Sorokin-Anton — APPROVED** [review]

File: src/jsonrpc2.ml [hunk 2]
```diff
@@ -82,13 +86,162 @@ module Make (IO : IO) : S with module IO = IO = struct
     let json = Jsonrpc.Notification.yojson_of_t m in
     send_json_ self json
 
+  let send_server_req (self : t) (m : Jsonrpc.Request.t) : unit IO.t =
```
[the remark this answers is not in the document]
Comment (heitor-lassarote) [reply src/jsonrpc2.ml:93]: Done.

[the remark this answers is not in the document]
Comment (heitor-lassarote) [reply src/jsonrpc2.ml:93]: This doesn't seem exposed, though, as it's not in the mli file.

[the remark this answers is not in the document]
Comment (Sorokin-Anton) [reply src/jsonrpc2.ml:93]: Oh, I see, there is a mli file which does not require changes

**Comment by heitor-lassarote** [comment]
@c-cube, seems like everything is fine and all approvals were given. Feel free to merge this whenever you want to. And let me know if you need anything else from me!

**Pull request merged**